<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>funki.pipelines &#8212; FUNKI documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/classic.css?v=d54626cc" />
    
    <script src="_static/documentation_options.js?v=22607128"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    
    <link rel="icon" href="_static/funki_favicon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="License" href="link_license.html" />
    <link rel="prev" title="funki.plots" href="plots.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="Related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="link_license.html" title="License"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="plots.html" title="funki.plots"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">FUNKI documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">funki.pipelines</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="module-funki.pipelines">
<span id="funki-pipelines"></span><h1>funki.pipelines<a class="headerlink" href="#module-funki.pipelines" title="Link to this heading">¶</a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="funki.pipelines.differential_expression">
<span class="sig-prename descclassname"><span class="pre">funki.pipelines.</span></span><span class="sig-name descname"><span class="pre">differential_expression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">design_factor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contrast_var</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_var</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logfc_thr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fdr_thr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pydeseq2'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#funki.pipelines.differential_expression" title="Link to this definition">¶</a></dt>
<dd><p>Computes differential expression analysis on the provided data based on a
given design factor and both the contrast and reference variables (e.g.
treatment and control). Then generates the resulting volcano plot based on
the desired thresholds.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<a class="reference internal" href="input.html#funki.input.DataSet" title="funki.input.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">funki.input.DataSet</span></code></a>) – The data from which to compute the differential expression</p></li>
<li><p><strong>design_factor</strong> (<em>str</em>) – Name of the column containing the variables which the
contrasting samples are assigned. The column must be present in the
<code class="docutils literal notranslate"><span class="pre">data.obs</span></code> table</p></li>
<li><p><strong>contrast_var</strong> (<em>any</em><em> | </em><em>list</em><em>[</em><em>any</em><em>]</em>) – The variable value(s) that defines the samples that are
to be contrasted against the reference (e.g. <code class="docutils literal notranslate"><span class="pre">'treatment'</span></code>). The value
must be present in the specified <code class="docutils literal notranslate"><span class="pre">design_factor</span></code> column</p></li>
<li><p><strong>ref_var</strong> (<em>any</em><em> | </em><em>list</em><em>[</em><em>any</em><em>]</em>) – The variable value(s) that defines the refence samples (e.g.
<code class="docutils literal notranslate"><span class="pre">'control'</span></code>). The value must be present in the specified
<code class="docutils literal notranslate"><span class="pre">design_factor</span></code> column</p></li>
<li><p><strong>logfc_thr</strong> (<em>float</em><em>, </em><em>optional</em>) – Threshold for signifacnce based on the log2(FC) value,
defaults to <code class="docutils literal notranslate"><span class="pre">1.0</span></code></p></li>
<li><p><strong>fdr_thr</strong> (<em>float</em><em>, </em><em>optional</em>) – Threshold for signifacnce based on the FDR value, defaults
to <code class="docutils literal notranslate"><span class="pre">0.05</span></code></p></li>
<li><p><strong>method</strong> (<em>str</em>) – Which method to use for computing the differential
expression. Available methods are <code class="docutils literal notranslate"><span class="pre">'pydeseq2'</span></code> or <code class="docutils literal notranslate"><span class="pre">'limma'</span></code>,
defaults to <code class="docutils literal notranslate"><span class="pre">'pydeseq2'</span></code>.</p></li>
<li><p><strong>ax</strong> (<a class="reference external" href="https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.html#matplotlib.axes.Axes">matplotlib.axes.Axes</a>) – Matplotlib Axes instance where to draw the plot. Defaults to
<code class="docutils literal notranslate"><span class="pre">None</span></code>, meaning a new figure and axes will be generated.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The figure contataining the resulting scatter plot. If an axes is
passed, nothing is returned.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://matplotlib.org/stable/api/_as_gen/matplotlib.figure.Figure.html#matplotlib.figure.Figure">matplotlib.figure.Figure</a> | None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="funki.pipelines.enrichment_analysis">
<span class="sig-prename descclassname"><span class="pre">funki.pipelines.</span></span><span class="sig-name descname"><span class="pre">enrichment_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">net</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contrast</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'source'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'target'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#funki.pipelines.enrichment_analysis" title="Link to this definition">¶</a></dt>
<dd><p>Performs enrichment analysis using <a class="reference external" href="https://decoupler-py.readthedocs.io/en/latest/index.html">Decoupler</a> based on a given network
(e.g. gene set collection) and statistical method(s) and returns a
figure with the consensus score across methods for the enrichment results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<a class="reference internal" href="input.html#funki.input.DataSet" title="funki.input.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">funki.input.DataSet</span></code></a>) – The data set from which to perform the enrichment</p></li>
<li><p><strong>net</strong> (<a class="reference external" href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html">pandas.DataFrame</a>) – The network linking the features of the data to the attributes
(e.g. pathways, gene sets, transcription factors, etc.)</p></li>
<li><p><strong>contrast</strong> (<em>str</em>) – Which result of the differential expression to use for the
enrichment. Must be present in <code class="docutils literal notranslate"><span class="pre">data.varm_keys</span></code> named with the format
<code class="docutils literal notranslate"><span class="pre">'{contrast_var}_vs_{ref_var}'</span></code>. Defaults to <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>method</strong> (<em>NoneType</em><em> | </em><em>str</em>) – Which statistical method to use in order to compute the
enrichment, defaults to <code class="docutils literal notranslate"><span class="pre">None</span></code>. If none is provided, uses <code class="docutils literal notranslate"><span class="pre">'ulm'</span></code>.
To see all the available methods, you can run <code class="docutils literal notranslate"><span class="pre">decoupler.mt.show()</span></code>
function.</p></li>
<li><p><strong>source</strong> (<em>str</em>) – Column name from the provided <code class="docutils literal notranslate"><span class="pre">net</span></code> containing the gene
sets to enrich for. Defaults to <code class="docutils literal notranslate"><span class="pre">'source'</span></code>.</p></li>
<li><p><strong>target</strong> (<em>str</em>) – Column name from the provided <code class="docutils literal notranslate"><span class="pre">net</span></code> containing the gene set
components (e.g. gene/protein names) that can be mapped back to the data
set variable names. Defaults to <code class="docutils literal notranslate"><span class="pre">'target'</span></code>.</p></li>
<li><p><strong>weight</strong> (<em>NoneType</em><em> | </em><em>str</em>) – Defines the column in the network containing the weights to
use in the enrichment, defaults to <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>top</strong> (<em>int</em>) – Number of top enriched gene sets to display based on their
consensus score. If a negative number is provided, the bottom ones will
be displayed instead.</p></li>
<li><p><strong>ax</strong> (<a class="reference external" href="https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.html#matplotlib.axes.Axes">matplotlib.axes.Axes</a>) – Matplotlib Axes instance where to draw the plot. Defaults to
<code class="docutils literal notranslate"><span class="pre">None</span></code>, meaning a new figure and axes will be generated.</p></li>
<li><p><strong>**kwargs</strong> (<em>optional</em>) – Other keyword arguments that are passed to the specific
method call from <a class="reference external" href="https://decoupler.readthedocs.io/en/latest/api/mt.html">decoupler.mt</a> methods</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">None</span></code>, results are stored inplace of the passed <code class="docutils literal notranslate"><span class="pre">data</span></code>
object, which is a <a class="reference internal" href="input.html#funki.input.DataSet" title="funki.input.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">funki.input.DataSet</span></code></a> instance. Estimates,
p-values and consensus scores (in case of multiple methods) are stored
as part of the <code class="docutils literal notranslate"><span class="pre">obsm</span></code> attribute of the object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>NoneType</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="funki.pipelines.sc_quality_control">
<span class="sig-prename descclassname"><span class="pre">funki.pipelines.</span></span><span class="sig-name descname"><span class="pre">sc_quality_control</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#funki.pipelines.sc_quality_control" title="Link to this definition">¶</a></dt>
<dd><p>Computes QC metrics on a single-cell data set and generates several plots
to visualize them. Generates a multipanel figure with the follwoing plots:</p>
<ul class="simple">
<li><p>Box plot with highest expression genes</p></li>
<li><p>Violin plot with number of genes per cell</p></li>
<li><p>Violin plot with total counts per gene</p></li>
<li><p>Violin plot with the percentage of mitochondrial genes per cell</p></li>
<li><p>Scatter plot of total counts vs. percentage of mitochondrial genes</p></li>
<li><p>Scatter plot of total counts vs. number of genes</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<a class="reference internal" href="input.html#funki.input.DataSet" title="funki.input.DataSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">funki.input.DataSet</span></code></a>) – The data set from which to compute the QC metrics</p></li>
<li><p><strong>ax</strong> (<a class="reference external" href="https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.html#matplotlib.axes.Axes">matplotlib.axes.Axes</a>) – Matplotlib Axes instance where to draw the plots. Defaults to
<code class="docutils literal notranslate"><span class="pre">None</span></code>, meaning a new figure and axes will be generated. If passed,
an axes with at least 2 columns and 3 rows is expected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The figure contataining the resulting plot with multiple panels
for different metrics and comparisons. If an axes is passed, nothing is
returned.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://matplotlib.org/stable/api/_as_gen/matplotlib.figure.Figure.html#matplotlib.figure.Figure">matplotlib.figure.Figure</a> | None</p>
</dd>
</dl>
</dd></dl>

</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
  <div>
    <h3><a href="index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">funki.pipelines</a><ul>
<li><a class="reference internal" href="#funki.pipelines.differential_expression"><code class="docutils literal notranslate"><span class="pre">differential_expression()</span></code></a></li>
<li><a class="reference internal" href="#funki.pipelines.enrichment_analysis"><code class="docutils literal notranslate"><span class="pre">enrichment_analysis()</span></code></a></li>
<li><a class="reference internal" href="#funki.pipelines.sc_quality_control"><code class="docutils literal notranslate"><span class="pre">sc_quality_control()</span></code></a></li>
</ul>
</li>
</ul>

  </div>
  <div>
    <h4>Previous topic</h4>
    <p class="topless"><a href="plots.html"
                          title="previous chapter">funki.plots</a></p>
  </div>
  <div>
    <h4>Next topic</h4>
    <p class="topless"><a href="link_license.html"
                          title="next chapter">License</a></p>
  </div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/pipelines.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="Related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="link_license.html" title="License"
             >next</a> |</li>
        <li class="right" >
          <a href="plots.html" title="funki.plots"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">FUNKI documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">funki.pipelines</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2024, Nicolàs Palacio-Escat.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    </div>
  </body>
</html>